### PR TITLE
Add in hashmod experimental plugin.

### PR DESCRIPTION
**Pull request by CrucisGamma**
Description of the Change
Add in the hashmod editor in as an experimental plugin.  Allows the user to specify a Key Column in a CSV file and then modify multiple attributes in the subsequent columns if the key attribute matches a row.

Alternate Designs
An option to have more than one key element would be good to add in (at the moment it's only 1).
I've also thought about having a chained amount of .csv files (pasted in as a file list) to allow the editor to apply multiple files worth of data at once.

Why Should This Be In Core?
This is supplying a way for users to self add to their own data with their own data sets. I can't see not wanting to have it available all the time (Geo.Country for example being a key for Geo.Latitude and Geo.Longitude is the most go to example I've come up with so far)

Benefits
A new way to set data automatically and controllable via an external CSV.

Possible Drawbacks
Nil

Verification Process
I was able to load Geo.Country data, verify it had no Geo.Latitude / Geo.Longitude and then used it to setup the Geo.Lat/Long and was able to use Scatter3D to then draw a crude map of the Mercator Projection of the earth in 3d.

Applicable Issues
Nil.

### CODE REVIEW

**Comment by CrucisGamma** [comment]
Due to issues with branches/forks etc I closed and reopened this (from this one https://github.com/constellation-app/constellation/pull/502).

File: CoreGraphUtilities/src/au/gov/asd/tac/constellation/graph/utilities/hashmod/Hashmod.java [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+ * limitations under the License.
+ */
+package au.gov.asd.tac.constellation.graph.utilities.hashmod;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.logging.Level;
+import static org.geotools.referencing.factory.ReferencingFactory.LOGGER;
```
Comment (antares1470) [inline CoreGraphUtilities/src/au/gov/asd/tac/constellation/graph/utilities/hashmod/Hashmod.java:23]: Use java.util.logging.Logger instead (shouldn't affect the logger code below)

File: CoreGraphUtilities/src/au/gov/asd/tac/constellation/graph/utilities/hashmod/Hashmod.java [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+        parser = new HashmodCSVImportFileParser();
+        try {
+            data = parser.parse(new HashmodInputSource(new File(csvFileStr)), null);
+        } catch (IOException ex) {
+            LOGGER.log(Level.SEVERE, ex.getLocalizedMessage(), ex);
+        }
+    }
+
+    public String[] getCSVFileHeaders() {
+        if (data != null && data.size() > 0) {
```
Comment (antares1470) [inline CoreGraphUtilities/src/au/gov/asd/tac/constellation/graph/utilities/hashmod/Hashmod.java:74]: replace `data.size() > 0` with `!data.isEmpty()`

File: CoreGraphUtilities/src/au/gov/asd/tac/constellation/graph/utilities/hashmod/Hashmod.java [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+            if (headers.length > 0) {
+                return headers[0];
+            }
+        }
+        return null;
+    }
+
+    public HashMap<String, Integer> getCSVKeys() {
+        final HashMap<String, Integer> keys = new HashMap<>();
+        if (data != null && data.size() > 0) {
```
Comment (antares1470) [inline CoreGraphUtilities/src/au/gov/asd/tac/constellation/graph/utilities/hashmod/Hashmod.java:99]: replace `data.size() > 0` with `!data.isEmpty()`

File: CoreGraphUtilities/src/au/gov/asd/tac/constellation/graph/utilities/hashmod/Hashmod.java [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+    public String getCSVHeader(final int col) {
+        final String[] headers = getCSVFileHeaders();
+        if (headers != null && headers.length > col) {
+            return headers[col];
+        }
+        return null;
+    }
+
+    public List<String[]> getCSVFileData() {
+        if (data != null && data.size() > 0) {
```
Comment (antares1470) [inline CoreGraphUtilities/src/au/gov/asd/tac/constellation/graph/utilities/hashmod/Hashmod.java:127]: replace `data.size() > 0` with `!data.isEmpty()`

File: CoreGraphUtilities/src/au/gov/asd/tac/constellation/graph/utilities/hashmod/Hashmod.java [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+            return data;
+        }
+        return null;
+    }
+
+    public String getValueFromKey(final String key, final int value) {
+        if (data != null && data.size() > 0) {
```
Comment (antares1470) [inline CoreGraphUtilities/src/au/gov/asd/tac/constellation/graph/utilities/hashmod/Hashmod.java:134]: replace `data.size() > 0` with `!data.isEmpty()`

File: CoreGraphUtilities/src/au/gov/asd/tac/constellation/graph/utilities/hashmod/Hashmod.java [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+            }
+        }
+        return null;
+    }
+
+    public Boolean doesKeyExist(final String key) {
+        if (key == null) {
+            return false;
+        }
+        if (data != null && data.size() > 0) {
```
Comment (antares1470) [inline CoreGraphUtilities/src/au/gov/asd/tac/constellation/graph/utilities/hashmod/Hashmod.java:152]: replace `data.size() > 0` with `!data.isEmpty()`

**Comment by spjspj** [comment]
Thanks @arcturus2 :)

**Comment by CrucisGamma** [comment]
Cool, thanks @arcturus2  :)

**Pull request merged**